src/main/java/de/tum/in/www1/artemis/service/ProgrammingExerciseService.java

```diff
@@ -992,6 +1009,32 @@ public void importBuildPlans(final ProgrammingExercise templateExercise, final P
         }
     }
 
+    /**
+     * Remove the write permissions for all students for their programming exercise repository.
+     * They will still be able to read the code, but won't be able to change it.
+     *
+     * @param programmingExerciseId     ProgrammingExercise id.
+     * @throws EntityNotFoundException  if the programming exercise can't be found.
+     */
+    public void removeWritePermissionsFromAllStudentRepositories(Long programmingExerciseId) throws EntityNotFoundException {
+        log.info("Invoking scheduled task programming exercise with id " + programmingExerciseId + ".");
+
+        ProgrammingExercise programmingExercise = findByIdWithEagerStudentParticipations(programmingExerciseId);
+        versionControlService.ifPresent(v -> {
+            for (StudentParticipation studentParticipation : programmingExercise.getStudentParticipations()) {
+                ProgrammingExerciseStudentParticipation programmingExerciseStudentParticipation = (ProgrammingExerciseStudentParticipation) studentParticipation;
+                try {
+                    v.setRepositoryPermissionsToReadOnly(programmingExerciseStudentParticipation.getRepositoryUrlAsUrl(), programmingExercise.getProjectKey(),
+                            programmingExerciseStudentParticipation.getStudent().getLogin());
+                }
+                catch (Exception e) {
+                    log.error("Removing write permissions failed for programming exercise with id " + programmingExerciseId + " for student repository with participation id "
```

Comment:
It think we should propagate these errors to the instructors in some way here since it's really critical that all repos get locked. Might be in a future PR.

Reply:
I agree. We don't have a UI concept for this yet, but we could in future e.g. send websocket messages to the client. I would agree that it is not in scope of this PR, but we should keep it in mind..

Reply:
After reading your review summary: I agree that an email notification (+ Artemis notification) would be best for this.

Reply:
I added at least a simple solution here:
We send a notification to the client that tells the instructor that:
a) All operations were successful
b) At least one operation was not succesful, and we also provide the number of failed operations here.

I prepared the implementation so that we could also send the concerned user names with the notification, but this could blow up the notification text, not sure if we want to do this. (notification text is longext).

Reply:
![image](https://user-images.githubusercontent.com/28230611/65973845-46fb8380-e46c-11e9-8de5-f3c3aa4bf409.png)